guvnor-ala/guvnor-ala-registry-local/src/main/java/org/guvnor/ala/registry/local/InMemoryRuntimeRegistry.java

```diff
@@ -0,0 +1,171 @@
+        providers.put( provider.getId(), provider );
+        if ( !providersByType.containsKey( provider.getProviderType() ) ) {
+            providersByType.put( provider.getProviderType(), new ArrayList<>() );
+        }
+        providersByType.get( provider.getProviderType() ).add( provider );
+    }
+
+    @Override
+    public List<Provider> getAllProviders() {
+        return new ArrayList<>( providers.values() );
+    }
+
+    @Override
+    public List<Provider> getProvidersByType( ProviderType type ) {
+        return providersByType.get( type );
+    }
+
+    @Override
+    public Provider getProvider( String providerName ) {
+        return providers.get( providerName );
+    }
+
+    @Override
+    public void unregisterProvider( Provider provider ) {
+        providersByType.get( provider.getProviderType() ).remove( provider );
+        providers.remove( provider.getId() );
+    }
+
+    @Override
+    public void unregisterProvider( String providerName ) {
+        for ( Provider p : providers.values() ) {
+            if ( p.getId().equals( providerName ) ) {
+                unregisterProvider( p );
+            }
+        }
+    }
+
+    @Override
+    public void registerRuntime( Runtime runtime ) {
+        final Provider provider = providers.get( runtime.getProviderId().getId() );
+
+        if ( !runtimesByProviderType.containsKey( provider.getProviderType() ) ) {
+            runtimesByProviderType.put( provider.getProviderType(), new ArrayList<>() );
+        }
+        runtimesByProviderType.get( provider.getProviderType() ).add( runtime );
+    }
+
+    @Override
+    public List<Runtime> getAllRuntimes() {
+        List<Runtime> runtimes = new ArrayList<>();
+        for ( List<Runtime> rs : runtimesByProviderType.values() ) {
+            runtimes.addAll( rs );
+        }
+        return runtimes;
+    }
+
+    @Override
+    public List<Runtime> getRuntimesByProvider( ProviderType providerType ) {
+        return new ArrayList<>( runtimesByProviderType.get( providerType ) );
+    }
+
+    @Override
+    public Runtime getRuntimeById( String id ) {
+        for ( ProviderType pt : runtimesByProviderType.keySet() ) {
+            for ( Runtime r : runtimesByProviderType.get( pt ) ) {
+                if ( r.getId().equals( id ) ) {
+                    return r;
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public <T extends Provider> Optional<T> getProvider( final ProviderId providerId,
+                                                         final Class<T> clazz ) {
+        checkNotNull( "providerId", providerId );
+        checkNotNull( "clazz", clazz );
+
+        final Provider value = providers.get( providerId.getId() );
+        if ( value == null ||
+                !clazz.isAssignableFrom( value.getClass() ) ) {
+            return Optional.empty();
+        }
+        return Optional.of( clazz.cast( value ) );
```

Comment:
There's nothing wrong here so feel free to ignore this :)

You can use `Optional` to simplify this into a "one-liner" that I think is still quite readable:

    return Optional.ofNullable( value )
        .filter( provider -> clazz.isInstance( provider ) )
        .map( provider -> clazz.cast( provider ) );

That evaluates to empty `value` is `null` or not an instance of `clazz`.